### PR TITLE
[ASTGen] Generate interpolated string literal

### PR DESCRIPTION
**Pull request by rintaro**
The first commit is just a cleanup in C++:
  * Remove unnecessary code from `parseExprStringLiteral()`/`parseStringSegments()` (leftover from libSyntax)
  * Remove unused `InterpolatedStringLiteralExpr::getTrailingQuoteLoc()` and `EditorPlaceholderExpr::getTrailingAngleBracketLoc()`. They were planned to be used in ASTScope, but not actually used.
  
For `InterpolatedStringLiteralExpr` generation, we probably should move the `TapExpr` construction to Sema (e.g. PrecheckExpression). But for now, ASTGen constructs the `TapExpr` and its interior.

### CODE REVIEW

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/2530
@swift-ci Please smoke test

File: test/ASTGen/exprs.swift [hunk 1]
```diff
@@ -103,3 +103,18 @@ func testTrailingClsure() {
...
+  _ = "\(arg)"
+  _ = """
+    foo
+    bar\
+    baz
+    """
+  _ = """
+    foo\(arg)
+    \(arg)bar
+    """
```
Comment (rintaro) [inline test/ASTGen/exprs.swift:119]: Add escaped characters tests.

**Review by hamishknight — APPROVED** [review]

File: test/ASTGen/exprs.swift (hunk 1 above)
Comment (hamishknight) [reply test/ASTGen/exprs.swift:119]: Could also add tests for things like string literals nested in interpolations (including multi-line literals) if you want

File: lib/ASTGen/Sources/ASTGen/Literals.swift [hunk 2]
```diff
@@ -11,18 +11,200 @@
...
+    }
+  }
 
-    // FIXME: Handle interpolated strings.
-    // FIXME: Avoid 'String' instantiation
-    var segment = node.segments.first!.as(StringSegmentSyntax.self)!.content.text
-    return segment.withBridgedString { bridgedSegment in
-      return .createParsed(self.ctx, value: bridgedSegment, loc: openDelimiterOrQuoteLoc)
+  func generateStringLiteralExpr(stringLiteralExpr node: StringLiteralExprSyntax) -> BridgedStringLiteralExpr {
+    let stringLiteralKind = node.stringLiteralKind ?? .singleLine
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/ASTGen/Sources/ASTGen/Literals.swift:27]: I think that’s not necessary. The parser will take care of that.

File: lib/ASTGen/Sources/ASTGen/Literals.swift [hunk 3]
```diff
@@ -11,18 +11,200 @@
...
+      return self.generateInterpolatedStringLiteralExpr(stringLiteralExpr: node).asExpr
+    }
+  }
 
-    // FIXME: Handle interpolated strings.
-    // FIXME: Avoid 'String' instantiation
-    var segment = node.segments.first!.as(StringSegmentSyntax.self)!.content.text
-    return segment.withBridgedString { bridgedSegment in
-      return .createParsed(self.ctx, value: bridgedSegment, loc: openDelimiterOrQuoteLoc)
+  func generateStringLiteralExpr(stringLiteralExpr node: StringLiteralExprSyntax) -> BridgedStringLiteralExpr {
```
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Literals.swift:26]: We should add a comment that this only handles the case where the string doesn’t contain any interpolation.

File: lib/ASTGen/Sources/ASTGen/Literals.swift [hunk 4]
```diff
@@ -11,18 +11,200 @@
...
+  func generateInterpolatedStringLiteralExpr(stringLiteralExpr node: StringLiteralExprSyntax) -> BridgedInterpolatedStringLiteralExpr {
+    let stringLiteralKind = node.stringLiteralKind ?? .singleLine
+    let delimiterLength = node.delimiterLength
+    let startLoc = self.generateSourceLoc(node)
+    let afterQuoteLoc: BridgedSourceLoc = {
+      var l = startLoc
+      if let pound = node.openingPounds {
+        l = l.advanced(by: pound.trimmedLength.utf8Length)
+      }
+      l = l.advanced(by: node.openingQuote.trimmedLength.utf8Length)
```
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Literals.swift:56]: This wouldn’t handle any trivia in between the pound and the opening quote. There shouldn’t be any, in theory but you never know how the parser might change. So, I think would be good to just add that as well.

**Review by ahoppen — APPROVED** [review]
Swift parts look good to me

File: lib/ASTGen/Sources/ASTGen/Literals.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/ASTGen/Sources/ASTGen/Literals.swift:27]: Should we be asserting then?

[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/ASTGen/Sources/ASTGen/Literals.swift:27]: Hmm, actually, on second thought. I think we should diagnose `nil` and `.singleQuote`. I didn’t realize `stringLiteralKind` was defined in swift-syntax.

[the remark this answers is not in the document]
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/Literals.swift:27]: `ParserDiagnostics` emits error for `.singleQuote` case,  so ASTGen should not, emit extra errors, IMO.
As for `nil` case, the SwiftParser probably won't construct trees like that (quote token kind is neither of `.stringQuote`, `.multilineStringQuote`, or `singleQuote`). But even if it does, I think it's parser's (or any tree generator's) responsibility to diagnose it.

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/2530
@swift-ci Please smoke test

File: lib/ASTGen/Sources/ASTGen/Literals.swift (hunk 3 above)
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/Literals.swift:26]: Added assertion too.

**Pull request merged**